test/utils/asserts.hpp

```diff
@@ -23,13 +25,57 @@ MATCHER(Result, "empty result")
 
 MATCHER_P(Result, value, "")  // NOLINT(readability-redundant-string-init)
 {
+    using namespace fizzy;
+
+    // Require the arg to be TypedExecutionResult.
+    // This can be a static_assert, but just returning false and failing a test provides better
+    // location of the error.
+    using result_type = std::remove_cv_t<std::remove_reference_t<arg_type>>;
+    if constexpr (!std::is_same_v<result_type, test::TypedExecutionResult>)
+    {
+        if (result_listener->IsInterested())
+            *result_listener << "TypedExecutionResult expected";
+        return false;
+    }
+
     if (arg.trapped || !arg.has_value)
         return false;
 
-    if constexpr (std::is_floating_point_v<value_type>)
-        return arg.value.template as<value_type>() == fizzy::test::FP{value};
-    else  // always check 64 bit of result for all integers, including 32-bit results
-        return arg.value.i64 == static_cast<std::make_unsigned_t<value_type>>(value);
+    if constexpr (std::is_same_v<result_type, test::TypedExecutionResult>)
+    {
+        if constexpr (std::is_same_v<value_type, float>)
+        {
+            return arg.type == ValType::f32 && arg.value.f32 == test::FP{value};
+        }
+
+        if constexpr (std::is_same_v<value_type, double>)
+        {
+            return arg.type == ValType::f64 && arg.value.f64 == test::FP{value};
+        }
+
+        if constexpr (std::is_integral_v<value_type> && sizeof(value_type) == sizeof(uint64_t))
+        {
+            return arg.type == ValType::i64 && arg.value.i64 == static_cast<uint64_t>(value);
+        }
+
+        if constexpr (std::is_integral_v<value_type> && sizeof(value_type) == sizeof(uint32_t))
+        {
+            if (arg.type == ValType::i32)
+            {
+                return arg.value.i64 == static_cast<uint32_t>(value);
+            }
+            else if (arg.type == ValType::i64 && value >= 0)
```

Comment:
Do you plan to remove this and change the tests at some point?

Reply:
Not sure. Probably make sense to always use `Result(0_u32)` or `Result(0_u64)` only.